sweepbatcher/presigned.go

```diff
@@ -0,0 +1,457 @@
+	for _, sweep := range b.sweeps {
+		sweeps = append(sweeps, sweep)
+	}
+
+	// Cache the destination address.
+	address, err := b.getSweepsDestAddr(ctx, sweeps)
+	if err != nil {
+		return 0, fmt.Errorf("failed to find destination address: %w",
+			err), false
+	}
+
+	// Construct unsigned batch transaction.
+	tx, weight, _, fee, err := constructUnsignedTx(
+		sweeps, address, currentHeight, feeRate,
+	)
+	if err != nil {
+		return 0, fmt.Errorf("failed to construct tx: %w", err),
+			false
+	}
+
+	// Adjust feeRate, because it may have been clamped.
+	feeRate = chainfee.NewSatPerKWeight(fee, weight)
+
+	// Calculate total input amount.
+	batchAmt := btcutil.Amount(0)
+	for _, sweep := range sweeps {
+		batchAmt += sweep.value
+	}
+
+	// Determine the current minimum relay fee based on our chain backend.
+	minRelayFee, err := b.wallet.MinRelayFee(ctx)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get minRelayFee: %w", err),
+			false
+	}
+
+	// Get a signed transaction. It may be either new transaction or a
+	// pre-signed one.
+	const presignedOnly = false
+	signedTx, err := b.cfg.presignedHelper.SignTx(
+		ctx, tx, batchAmt, minRelayFee, feeRate, presignedOnly,
+	)
+	if err != nil {
+		return 0, fmt.Errorf("failed to sign tx: %w", err),
+			false
+	}
+
+	// Run sanity checks to make sure presignedHelper.SignTx complied with
+	// all the invariants.
+	err = CheckSignedTx(tx, signedTx, batchAmt, minRelayFee)
+	if err != nil {
+		return 0, fmt.Errorf("signed tx doesn't correspond the "+
+			"unsigned tx: %w", err), false
+	}
+	tx = signedTx
+	txHash := tx.TxHash()
+
+	// Make sure tx weight matches the expected value.
+	realWeight := lntypes.WeightUnit(
+		blockchain.GetTransactionWeight(btcutil.NewTx(tx)),
+	)
+	if realWeight != weight {
+		b.Warnf("actual weight of tx %v is %v, estimated as %d",
+			txHash, realWeight, weight)
+	}
+
+	// Find actual fee rate of the signed transaction. It may differ from
+	// the desired fee rate, because SignTx may return a presigned tx.
+	output := btcutil.Amount(tx.TxOut[0].Value)
+	fee = batchAmt - output
+	signedFeeRate := chainfee.NewSatPerKWeight(fee, realWeight)
+
+	numSweeps := len(tx.TxIn)
+	b.Infof("attempting to publish custom signed tx=%v, desiredFeerate=%v,"+
+		" signedFeeRate=%v, weight=%v, fee=%v, sweeps=%d, destAddr=%s",
+		txHash, feeRate, signedFeeRate, weight, fee, numSweeps, address)
+	b.debugLogTx("serialized batch", tx)
+
+	// Publish the transaction.
+	err = b.wallet.PublishTransaction(ctx, tx, b.cfg.txLabeler(b.id))
+	if err != nil {
+		return 0, fmt.Errorf("publishing tx failed: %w", err), true
+	}
+
+	// Store the batch transaction's txid and pkScript, for monitoring
+	// purposes.
+	b.batchTxid = &txHash
+	b.batchPkScript = tx.TxOut[0].PkScript
+
+	return fee, nil, true
+}
+
+// getSweepsDestAddr returns the destination address used by a group of sweeps.
+// The method must be used in presigned mode only.
+func (b *batch) getSweepsDestAddr(ctx context.Context,
```

Comment:
could call it `getPresignedSweepsDestAddr` to make it clearer that this only makes sense in the context of presigned batches.

Reply:
Renamed